Data/ByteString/Builder.hs

```diff
@@ -440,18 +442,20 @@ char8 :: Char -> Builder
 char8 = P.primFixed P.char8
 
 -- | Char8 encode a 'String'.
-{-# INLINE [1] string8 #-} -- phased to allow P.cstring rewrite
+{-# INLINE [1] string8 #-} -- phased to allow literal cstring rewrites
```

Comment:
@chessai , @Bodigrim , @clyring A question for the reviewers:

Why is the phase specified here equal to `1`?  When I add tests to see whether `string8` and `stringUtf8` actually benefit from the RULES, I only get improvement when the phase is set to `0`:

```haskell
--- a/bench/BenchAll.hs
+++ b/bench/BenchAll.hs
@@ -255,6 +255,10 @@ ascBuf, utfBuf :: Ptr Word8
 ascBuf = Ptr "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"#
 utfBuf = Ptr "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\xc0\x80xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"#
 
+ascStr, utfStr :: String
+ascStr = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+utfStr = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\0xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+
 asclit, utflit :: Ptr Word8 -> Builder
 asclit str@(Ptr addr) = BI.ascLiteralCopy str (byteCountLiteral addr)
 utflit str@(Ptr addr) = BI.modUtf8LitCopy str (byteCountLiteral addr)
@@ -273,6 +277,8 @@ main = do
         , benchB' "String"        () $ \() -> asclit (Ptr "hello world!"#)
         , benchB' "AsciiLit"      () $ \() -> asclit ascBuf
         , benchB' "Utf8Lit"       () $ \() -> utflit utfBuf
+        , benchB' "strLit"        () $ \() -> string8 ascStr
+        , benchB' "utfLit"        () $ \() -> stringUtf8 utfStr
         ]
 
       , bgroup "Encoding wrappers"
```

With the phase set to `1` as a baseline, testing with phase `0` the `bench` report is:
```
All
  Data.ByteString.Builder
    Small payload
      AsciiLit: OK (2.64s)
        243  ns ± 8.5 ns,       same as baseline
      Utf8Lit:  OK (1.45s)
        286  ns ±  15 ns,       same as baseline
      strLit:   OK (1.23s)
        243  ns ±  19 ns, 51% less than baseline
      utfLit:   OK (1.38s)
        279  ns ±  15 ns, 44% less than baseline
```

This is with GHC 9.4.5.

Reply:
Testing with GHC 8.10 the phase 1 -> phase 0 delta is:
```
All
  Data.ByteString.Builder
    Small payload
      AsciiLit: OK (1.00s)
        179  ns ± 9.5 ns, 47% less than baseline
      Utf8Lit:  OK (0.53s)
        198  ns ±  15 ns,       same as baseline
      strLit:   OK (0.49s)
        187  ns ±  19 ns, 51% less than baseline
      utfLit:   OK (0.52s)
        199  ns ±  20 ns,       same as baseline
```
With GHC 9.2:
```
All
  Data.ByteString.Builder
    Small payload
      AsciiLit: OK (1.36s)
        236  ns ±  22 ns,       same as baseline
      Utf8Lit:  OK (1.36s)
        274  ns ±  16 ns,       same as baseline
      strLit:   OK (1.20s)
        237  ns ±  14 ns, 69% less than baseline
      utfLit:   OK (1.36s)
        275  ns ±  15 ns, 64% less than baseline
```

Reply:
I'm not sure. In principle these rules can fire in phase 2, and I do observe "Rule fired: string8/unpackFoldrCString# (Data.ByteString.Builder)" if I inline ascStr in your example and compile with ghc-9.4.4.

These should probably just be NOINLINE pragmas. primMapList{Fixed,Bounded} are themselves marked inline (to encourage good specialization with the particular `BoundedPrim` used) and produce lots of code, but will not actually fuse with good list producers.

Reply:
In HEAD (master), with no changes other than phase [1] -> [0]:
```
All
  Data.ByteString.Builder
    Small payload
      strLit: OK (1.90s)
        688  ns ±  43 ns, 27% less than baseline
      utfLit: OK (0.93s)
        736  ns ±  62 ns, 31% less than baseline
```
Which seems to suggest that the original phase control was not helping, indeed simply removing the rules and inlining (GHC 9.2) gives:
```
All
  Data.ByteString.Builder
    Small payload
      strLit: OK (1.14s)
        790  ns ±  52 ns, 16% less than baseline
      utfLit: OK (0.99s)
        791  ns ±  66 ns, 26% less than baseline

All 2 tests passed (2.18s)
```

Reply:
Could the issue be that I'm giving `string8` and `stringUtf8` named constant strings, rather than inline string constants? That is, first inline the arguments at the call site, and only then inline `string8`?

This looks rather fragile. Is there a downside to setting the phase number to `0`?

Reply:
It is indeed fragile, and it doesn't even work with a non-ASCII string. (Looking at the core2core output, it looks like the `unpackCStringUtf8#` gets rewritten in terms of `unpackFoldrCStringUtf8#` before our rule fires...) Setting the phase number to 0 might help, too, but my suggestion was more extreme:

```suggestion
{-# NOINLINE string8 #-} -- allow literal cstring rewrites
```

Reply:
> It is indeed fragile, and it doesn't even work with a non-ASCII string. (Looking at the core2core output, it looks like the `unpackCStringUtf8#` gets rewritten in terms of `unpackFoldrCStringUtf8#` before our rule fires...) Setting the phase number to 0 might help, too, but my suggestion was more extreme:

I am beginning to agree.  And I don't think this impairs the efficiency of the non-literal input case, where the particular BoundedPrim is already available, and that's all that needs to be optimised, the input data does not have be seen for good code generation.  And rewrite RULES fire without inlining the result, so this makes sense. I'll push a commit with `NOINLINE` and the additional benchmark variants.